src/node_file.cc

```diff
@@ -474,14 +481,12 @@ static void InternalModuleReadJSON(const FunctionCallbackInfo<Value>& args) {
     numchars = uv_fs_read(loop, &read_req, fd, &buf, 1, offset, nullptr);
     uv_fs_req_cleanup(&read_req);
 
-    CHECK_GE(numchars, 0);
+    if (numchars < 0)
+      return;
```

Comment:
please return the empty string so that it gets put into `packageMainCache`

Reply:
That doesn't sound right.  `''` is for a package.json without a `main` property, see the bottom of this function.

Reply:
I'm look at https://github.com/nodejs/node/blob/e1c29f2c529ffdbf9cf8f05d4ed27ccfcede2719/lib/module.js#L133 which eagerly returns with a falsey value and sets an entry. https://github.com/nodejs/node/blob/e1c29f2c529ffdbf9cf8f05d4ed27ccfcede2719/lib/module.js#L148 then bails out. this also looks like it might have a bug re-reading it.

Reply:
Yes, but that's for when we found (and read) a package.json.  That's not the case here, it's closer to an ENOENT or EPERM condition.

Reply:
I disagree, we found a `package.json` it was a directory though.

Reply:
Exactly, not a file we can parse.

Reply:
The intent of this change is around `package.json/` directories. This mimics a `package.json` that was found but has no relevant `"main"` entry : https://github.com/bnoordhuis/io.js/blob/ec93af3bd79c89088f832f72d788225abb106910/src/node_file.cc#L496-L497 . We can discern the `"main"` we want to represent for it without the need to parse, and returning `""` causes the `packageMainCache` to correctly store the falsey value to bail the module system.

Reply:
I'm kind of done arguing but I'll address your last point: by your logic errors such as EPERM should also be interpreted as meaning "package.json without a main entry" - but we don't do that and rightly so.

Reply:
EPERM is different you can EPERM a directory. We can inspect this entity on the file system.